src/utils/config_filters.cpp

```diff
@@ -121,3 +121,144 @@ bool utils::config_filters::bool_or_empty(const config& filter, const config& cf
 	}
 	return is_matches;
 }
+
+bool utils::config_filters::matches_ability_filter(const config & cfg, const std::string& tag_name, const config & filter)
+{
+
+	if(!filter["affect_adjacent"].empty()){
+		bool adjacent = cfg.has_child("affect_adjacent");
+		if(filter["affect_adjacent"].to_bool() != adjacent){
+			return false;
+		}
+	}
+
+	if(!bool_matches_if_present(filter, cfg, "affect_self", true))
+		return false;
+
+	if(!bool_or_empty(filter, cfg, "affect_allies"))
+		return false;
+
+	if(!bool_matches_if_present(filter, cfg, "affect_enemies", false))
+		return false;
+
+	if(!bool_matches_if_present(filter, cfg, "cumulative", false))
+		return false;
+
+	const std::vector<std::string> filter_type = utils::split(filter["tag_name"]);
+	if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), tag_name) == filter_type.end() )
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "overwrite_specials", "none"))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "id", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "apply_to", "self"))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "active_on", "both"))
+		return false;
+
+	//for damage_type only
+	if(!string_matches_if_present(filter, cfg, "replacement_type", ""))
+		return false;
+
+	if(!string_matches_if_present(filter, cfg, "alternative_type", ""))
+		return false;
+
+	//for plague only
+	if(!string_matches_if_present(filter, cfg, "type", ""))
+		return false;
+
+	if(exclude_tags_.count(tag_name) == 0){
+		if(!filter["value"].empty()){
+			if(tag_name == "drains"){
+				if(!int_matches_if_present(filter, cfg, "value", 50)){
+					return false;
+				}
+			} else if(tag_name == "berserk"){
+				if(!int_matches_if_present(filter, cfg, "value", 1)){
+					return false;
+				}
+			} else if(tag_name == "heal_on_hit" || tag_name == "heals" || tag_name == "regenerate" || tag_name == "leadership"){
+				if(!int_matches_if_present(filter, cfg, "value" , 0)){
+					return false;
+				}
+			} else {
+				if(!int_matches_if_present(filter, cfg, "value")){
+					return false;
+				}
+			}
+		}
+		if(tag_name != "berserk"){
```

Comment:
Same as above. This makes it really complicated to figure out what tags this checks. Some comments would help.

Reply:
comments added